kdc/kerberos5.c

```diff
@@ -2115,8 +2286,23 @@ _kdc_as_rep(astgs_request_t r)
 	goto out;
     }
 
-    ret = _krb5_principalname2krb5_principal(r->context, &r->client_princ,
```

Comment:
Heimdal _does_ support PKINIT SANs (I use that all the time).  So you don't have to write code to extract the SANs from the client's PKINIT certificate.  What Heimdal doesn't support is X.500-style principal names.  So I think what we need here is to inspect the `cname` from the `AS-REQ` and decide that what we should do is take the Kerberos principal name from the certificate (if it's present there).

But I think too X.500 names could be checked against the DN of the certificate if it doesn't have a PKINIT SAN, and then maybe we issue the ticket with the requested name (but, uh, no Heimdal or MIT services will understand what to do with it).

Reply:
> So you don't have to write code to extract the SANs from the client's PKINIT certificate.

The reason I duplicated the code to extract the SAN is because the PADATA parsing and validation code happens after the "AS-REQ_cname to principal_name" code. I initially tried re-using the same PA-DATA extraction code earlier, but it depends on the client_principal to already have been fetched from the database. I'm new to this code though, so is there an easier way of grabbing the client_cert from the PADATA earlier than the "cname to principal" code? I am already reusing functions to extract the SAN from the client PKINIT certificate, but I'm just not able to find an easy way to get the client_cert struct in the first place.

> But I think too X.500 names could be checked against the DN of the certificate if it doesn't have a PKINIT SAN, and then maybe we issue the ticket with the requested name

That was my thought. As a first step (since it's already current behaviour right now), we can prioritize the SAN as the principal name (if it exists), and otherwise fallback to using the cname verbatim. Then in future commits, we can allow a configuration option to run a regexp pattern match against the cert's DN and extract whatever key=value pair as the principal name.

Reply:
> I'm new to this code though,

And because of that I'm quite impressed with your PR, so please don't be discouraged.

> so is there an easier way of grabbing the client_cert from the PADATA earlier than the "cname to principal" code?

I see your problem now, yes: we look up the requested client principal name way too early to hook this into the existing pre-auth code.

So, yeah, I think we may want to go with roughly what you sketched.  Though we'll want to check whether the requested `cname` is of X.500 name-type and only then do this check.  By the way, this is a good time to point out that we do support using PKINIT w/ PKINIT SANs in certificates even when the named principals don't exist in the HDB -- all you have to do is enable synthetic principals (currently set `synthetic_clients = true` in `[kdc]` in `krb5.conf`, though I'll be moving that into `[hdb]` soon).

So an alternative sketch would be to a) parse X.500 names, b) treat them as "synthetic" so we can get past the early `cname` lookup in the HDB, c) later in the PKINIT code (specifically in `_kdc_pk_check_client()`) replace the requested `cname` with the first PKINIT SAN in the client's certificate.  Now, this does seem a bit hackish, but so does trying to extract the SAN very early, so I'd be happy with either approach, and this alternative would be harder for you to write, so let's go with your approach.

Reply:
So I guess what I'd like to see here is a change where we call `siva_extract_san_from_padata()` only if `b->cname` is of X.500 name-type :)

Reply:
Actually turns out after further testing, none of this is needed. On windows, only the first AS-REQ is made with the NT-X500-PRINCIPAL, and it contains no PADATA. I have some local code to simply run a POSIX regex against a specific attribute in the subject DN (common name seemed like a good start) and extract the value as the principal name. This sent back an ERR_PREAUTH_REQUIRED to Windows, which then sent a subsequent AS-REQ with PADATA and a normal principal entry. The Heimdal KDC returned back a successful AS-REP.

However, windows still doesn't want to consider that as a successful login attempt for some reason, so I'm looking into it further. I'll update this draft PR soon with what I have. In my local changes, I am only running this when the cname is of type NT-X500-PRINCIPAL.